File: 13_linked_list_find/findValue.js
// class Node {
//   constructor(val) {
//     this.val = val;
//     this.next = null;
//   }
// }

// const linkedListFind = (head, target) => {
//   let currentNode = head
//   while(currentNode){
//      if(currentNode.val === target){
//     return true
//   }
//     currentNode = currentNode.next
//   }
// return false
// };

const linkedListFind = (head, target) => {
  if(!head){
    return false
  }
  else if(head.val === target){
    return true
  }
  else{
    return linkedListFind(head.next, target)
  }
}

// Alvin's Solutions

// const linkedListFind = (head, target) => {
//   let current = head;
//   while (current !== null) {
//     if (current.val === target) return true;
//     current = current.next;
//   }
//   return false;
// };

// const linkedListFind = (head, target) => {
//   if (head === null) return false;
//   if (head.val === target) return true;
//   return linkedListFind(head.next, target);
// };

File: Structy_1/03_Linked_List/13_linked_list_find/find_value.py
class Node:
  def __init__(self, val):
    self.val = val
    self.next = None
    
def linked_list_find(head, target):
  if head is None:
    return False
  if head.val == target:
    return True
  return linked_list_find(head.next, target)

# def linked_list_find(head, target):
#   current_node = head
#   while current_node is not None:
#     if current_node.val == target:
#       return True
#     current_node = current_node.next
#   return False

# Both have O(n), but iterative uses constant space and recursive uses linear because of callstack creations

# Alvin's Solutions

def linked_list_find(head, target):
  current = head
  while current is not None:
    if current.val == target:
      return True
    current = current.next
  return False

def linked_list_find(head, target):
  if head is None:
    return False
  if head.val == target:
    return True
  return linked_list_find(head.next, target)